### PR TITLE
punycode: use WPACKET for building the result

### PR DESCRIPTION
**Pull request by paulidale**
Removing the recently added bounds checking code in favour of our WPACKET code.

- Added a unit test that fails before the recent CVE fix & passes with it.  This PR isn't changing the behaviour per say.
- Added a notional number of bytes that were attempted to be written to WPACKET, so the required space return is possible.
- Converted the punycode code to use WPACKET.
- Added a fuzzer

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by paulidale** [comment]
We might want to cherry pick the x509 commit to 3.0.
Otherwise, we're accepting strings one character fewer than previously.

File: crypto/packet.c
Comment (t8m) [inline crypto/packet.c:491]: This is IMO unnecessary. Instead use the WPACKET_init_null() for the NULL destination buffer to get the number of bytes needed in the buffer.

File: crypto/punycode.c
Comment (t8m) [inline crypto/punycode.c:276]: We can change the semantics of ossl_a2ulabel to count the necessary bytes only in the out==NULL case. Also in that case we could make it return success as there is no real failure. The current semantics is quite awkward.

File: crypto/packet.c
Comment (paulidale) [reply crypto/packet.c:491]: This would work for a NULL destination however the caller also needs to return the correct length if the writes overflow.  I'd need to write everything twice in that case.

File: crypto/punycode.c
Comment (paulidale) [reply crypto/punycode.c:276]: Exactly.....

It's never called with a NULL output buffer.

Maybe this entire thing needs a rethink.  The returned size doesn't appear to be used so maybe the length return can be removed entirely.

File: crypto/punycode.c
Comment (beldmit) [inline crypto/punycode.c:248]: If I remember correctly, the logic I intended here was the outlen is the return value - e.g. if passed buffer is not enough, we could reallocate it and repeat the operation. But I may be wrong.

Comment (paulidale) [reply crypto/punycode.c:248]: Yeah, that was what was implemented but it's nasty to maintain such semantics using WPACKET.

The alternative would be to keep the returned length but only return the required size if `out == NULL`.  For `out != NULL` return the used size or that the buffer wasn't large enough (with no indication as to how large it needs to be).  The calling would then be:

```c
ossl_a2ulabel(in, NULL, &size);
out = malloc(size);
ossl_a2ulabel(in, out, &size);
```

The two places this function is called don't check the returned size, so I figured it was better to simplify the code and not worry about the `out == NULL` case or return the required size on overflow.

I can change things if it's deemed desirable but, IMO, it would be best if we required the functionality first.

File: test/punycode_test.c
Comment (t8m) [inline test/punycode_test.c:203]: Nit: `dotty` makes one think it has something with terminal devices. Could you please choose a different name?

Comment (paulidale) [reply test/punycode_test.c:203]: Sigh....

File: crypto/punycode.c
Comment (t8m) [reply crypto/punycode.c:248]: Yes, but we do not need this in the current code. If we ever need this, we can add it. This is internal thing.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Approved assuming CI passes.

**Comment by paulidale** [comment]
In a separate PR ?

**Comment by thesamesam** [comment]
I'd go for a separate PR as some tinkering might be needed to make CI happy.

**Comment by mspncp** [comment]
> In a separate PR ?

A separate PR is ok, as long as it happens. We should not make the same mistake twice.

File: fuzz/punycode.c
Comment (mspncp) [inline fuzz/punycode.c:34]: nit: indentation

**Comment by paulidale** [comment]
I've added some fuzzing.  If the tests are happy, I'll leave it.  If not I'll move it to a new PR.

**Comment by paulidale** [comment]
The fuzzing isn't so trivial as that post made out....
Still, it's found a memory leak already.

**Comment by thesamesam** [comment]
> The fuzzing isn't so trivial as that post made out.... Still, it's found a memory leak already.

It's motivated me to try out libfuzzer at some point, as it supposedly makes such things easier.

File: fuzz/punycode.c
Comment (thesamesam) [inline fuzz/punycode.c:11]: This is repeated

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (thesamesam) [inline doc/internal/man3/ossl_punycode_decode.pod:40]:
```suggestion
-1 when an invalid PUNYCODE string is passed.
```

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (thesamesam) [inline doc/internal/man3/ossl_punycode_decode.pod:37]:
```suggestion
-1 if an invalid PUNYCODE string is passed.
```

File: fuzz/punycode.c
Comment (kroeckx) [inline fuzz/punycode.c:33]: Why doesn't this just pass buf? I see no reason to alloc b with an extra \0.

Comment (kroeckx) [reply fuzz/punycode.c:33]: It seems to be needed for ossl_a2ulabel, but not for ossl_punycode_decode

Comment (paulidale) [reply fuzz/punycode.c:33]: The fuzzer doesn't \0 terminate what it passes in and `ossl_a2ulabel()` on the next line crashes if there is no terminator.
I could pass either to `ossl_punycode_decode()` since they are identical but I do need to alloc b and add the \0 for the next line.

Comment (kroeckx) [reply fuzz/punycode.c:33]: I prefer the call to ossl_punycode_decode be the not \0 terminated version.

Comment (paulidale) [reply fuzz/punycode.c]: Done FWIW.

**Comment by t8m** [comment]
> The fuzzing isn't so trivial as that post made out.... Still, it's found a memory leak already.

Hmm... this made me to realize that WPACKET always does memory allocation internally even for use-cases where it is not needed. Perhaps it is really heavier than necessary for many use-cases.

**Comment by paulidale** [comment]
I did mention the weight of WPACKET and my concerns were dismissed out of hand.  Hence this PR.
Can't have it both ways.  Either it's heavy or it isn't.

**Comment by t8m** [comment]
> I did mention the weight of WPACKET and my concerns were dismissed out of hand. Hence this PR.
> Can't have it both ways. Either it's heavy or it isn't.

I've changed my mind - it is a little bit heavy - could we make it leaner?

**Comment by paulidale** [comment]
I'd need to write an overflow safe buffer library that doesn't do any memory allocation (i.e. it isn't WPACKET but something new).  I'd leave out the sub packet processing etc.  Essentially it would be append only with memcpy, strcpy, put byte, get length & probably a few others.

The punycode decoder requires an insert operations and I'm not currently doing this with WPACKET because it's not supported.  Bad range checking in this code was one of the CVEs 😞

**Comment by slontis** [comment]
> I did mention the weight of WPACKET and my concerns were dismissed out of hand. Hence this PR. Can't have it both ways. Either it's heavy or it isn't.

You did not explain before why it was heavy..

**Comment by slontis** [comment]
> I've added some fuzzing. If the tests are happy, I'll leave it. If not I'll move it to a new PR.

Can you do a quick test of this against the old dodgy code as well?

**Comment by paulidale** [comment]
The fuzzing does crash the old code.  See the two corpora I already included.

**Comment by paulidale** [comment]
> You did not explain before why it was heavy..

I rather suspect this is getting the onus backwards.

**Comment by paulidale** [comment]
This has one zalloc overhead.  Making it leaner is going to require quite a bit of effort.  I'm not convinced it worthwhile especially given the heaviness of certificate processing where it occurs.

**Comment by slontis** [comment]
If it was added I would expect that WPACKET might use the lightweight layer. (Having any type of error in the new code would not be a good thing though :))

**Comment by thesamesam** [comment]
Is it worth trying to do some benchmarks and see if there's an actual problem? Some of the previous problems have been because of determination to increase performance but that's not always the right tradeoff (see e.g. the AVX512 discussion we're having).

Could also maybe let it sit in master and see how things go, but not backport it.

**Comment by t8m** [comment]
> Could also maybe let it sit in master and see how things go, but not backport it.

There is currently no intention to backport it to 3.0 as this is a refactoring not a bug fix.

**Comment by t8m** [comment]
> This has one zalloc overhead. Making it leaner is going to require quite a bit of effort. I'm not convinced it worthwhile especially given the heaviness of certificate processing where it occurs.

This surely can be merged without changing WPACKET. This is not performance critical peace of code anyway. It's just that if we want to use WPACKET truly everywhere we should first consider to make it leaner for the cases where the output buffer is static and there is no need for subpacket support.

**Comment by paulidale** [comment]
It needs approvals before it can merge...

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (t8m) [inline doc/internal/man3/ossl_punycode_decode.pod:37]: .... or on other errors.

File: doc/internal/man3/ossl_punycode_decode.pod
Comment (t8m) [inline doc/internal/man3/ossl_punycode_decode.pod:40]: same here

**Review by t8m — CHANGES_REQUESTED** [review]
Just doc nits

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Ping for second review.

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
merged.

**Pull request closed without merging**
